tests/wlr_layer_shell_v1.cpp

```diff
@@ -0,0 +1,564 @@
+/*
+ * Copyright © 2019 Canonical Ltd.
+ *
+ * This program is free software: you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 3,
+ * as published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ * Authored by: William Wold <william.wold@canonical.com>
+ */
+
+#include "helpers.h"
+#include "in_process_server.h"
+#include "layer_shell_v1.h"
+#include "xdg_shell_stable.h"
+
+#include <gmock/gmock.h>
+
+using namespace testing;
+
+using Vec2 = std::pair<int, int>;
+using Rect = std::pair<Vec2, Vec2>;
+
+namespace
+{
+
+class LayerShellTest
+    : public wlcs::StartedInProcessServer
+{
+private:
+    wlcs::CheckInterfaceExpected expect_layer_shell_supported{the_server(), zwlr_layer_shell_v1_interface};
+};
+
+class LayerSurfaceTest: public wlcs::StartedInProcessServer
+{
+public:
+    LayerSurfaceTest()
+        : client{the_server()},
+          surface{client},
+          layer_surface{client, surface}
+    {
+    }
+
+    void commit_and_wait_for_configure()
+    {
+        wl_surface_commit(surface);
+        layer_surface.dispatch_until_configure();
+    }
+
+    void expect_surface_is_at_position(std::pair<int, int> pos)
+    {
+        auto pointer = the_server().create_pointer();
+        pointer.move_to(pos.first + 2, pos.second + 3);
+        client.roundtrip();
+
+        ASSERT_THAT(client.focused_window(), Eq((wl_surface*)surface));
+        ASSERT_THAT(wl_fixed_to_int(client.pointer_position().first), Eq(2));
+        ASSERT_THAT(wl_fixed_to_int(client.pointer_position().second), Eq(3));
```

Comment:
I generally prefer `EXPECT_THAT` over `ASSERT_THAT` for expectations that are not preconditions of later expectations.

In this case, `x != 2` does not mean we can't check the value of `y`, so these should be `EXPECT_THAT`, right?

Reply:
Agreed